### PR TITLE
[SERF-1710] Metrics and logs for Kafka client

### PR DESCRIPTION
**Pull request by Neurostep**
## Description

In this PR we introduce metrics and logs for the Kafka client. Our current Kafka client `franz-go` provides a rich API for plugin publishing metrics by [writing Hooks](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#Hook).

We split up the metrics into 3 groups: `broker metrics`, `producer metrics`, and `consumer metrics`.

Potentially, adding metrics via hooks can lead to slow high rate producers/consumers, we provide a way to opt-in/opt-out those 3 groups.

`franz-go` also provides an [API to plugin logger](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#Logger). In this PR we provide a plugin for `franz-go` and use `go-sdk` logger.

## Testing considerations

- [x] Deploy one of the chassis applications and check metrics published
- [x] Deploy one of the chassis applications and check logs

## Checklist

- [x] Prefixed the PR title with the JIRA ticket code 
- [x] Performed simple, atomic commits with [good commit messages][commit messages]
- [x] Verified that the commit history is linear and commits are squashed as necessary
- [x] Thoroughly tested the changes in `development` and/or `staging`
- [x] Updated the `README.md` as necessary

## Related links

* [SERF-1710](https://scribdjira.atlassian.net/browse/SERF-1710)

[commit messages]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: pkg/metrics/kafka/metrics.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+	return &BrokerMetrics{metrics: m, commonMetrics: commonMetrics{metrics: m}}
+}
+
+func (b *BrokerMetrics) OnBrokerThrottle(
+	meta kgo.BrokerMetadata,
+	throttleInterval time.Duration,
+	throttledAfterResponse bool) {
+	node := strconv.Itoa(int(meta.NodeID))
+	b.commonMetrics.Histogram(
+		"kafka_client.broker.throttle_latencies",
```
[the remark this answers is not in the document]
Comment (Neurostep) [reply pkg/metrics/kafka/metrics.go]: @terranisu nice one 👍 Updated the metrics name ✅

File: README.md [hunk 2]
```diff
@@ -841,23 +841,26 @@ Kafka top-level configuration contains generic options that fit both publisher a
 | Cert PEM Key             | Client's private key string (PEM format) used for authentication  | `cert_pem_key`             | `APP_PUBSUB_KAFKA_CERT_PEM_KEY`             | string       | long PEM key string (deprecated)                      |
 | Security Protocol        | Protocol used to communicate with brokers                         | `security_protocol`        | `APP_PUBSUB_KAFKA_SECURITY_PROTOCOL`        | string       | plaintext, ssl, sasl_plaintext, sasl_ssl (deprecated) |
 | SSL verification enabled | Enable OpenSSL's builtin broker (server) certificate verification | `ssl_verification_enabled` | `APP_PUBSUB_KAFKA_SSL_VERIFICATION_ENABLED` | bool         | true, false (deprecated)                              |
+| Enable Metrics           | Enable publishing metrics for Kafka client broker connection      | `metrics_enabled`          | `APP_PUBSUB_KAFKA_METRICS_ENABLED`          | bool         | true, false                                           |
```
Comment (fotos) [inline README.md:844]: Nit suggestion:k

```suggestion
| Enable Metrics           | Enable publishing metrics for Kafka client broker connection      | `metrics_enabled`          | `APP_PUBSUB_KAFKA_BROKER_METRICS_ENABLED`          | bool         | true, false                                           |
```

Maybe adding `BROKER` makes it more clear what this setting is about? 🤔

File: pkg/logger/kafka/kafka.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	switch level {
+	case logger.Panic, logger.Error, logger.Fatal:
+		kgoLevel = kgo.LogLevelError
+	case logger.Warn:
+		kgoLevel = kgo.LogLevelWarn
+	case logger.Info:
+		kgoLevel = kgo.LogLevelInfo
+	case logger.Trace, logger.Debug:
+		kgoLevel = kgo.LogLevelDebug
+	}
```
Comment (fotos) [inline pkg/logger/kafka/kafka.go:58]: I'm not sure but do we need a default case? 🤔 

```golang
default:
  kgoLevel = kgo.LogLevelDebug
```

Or is `LogLevelNone` a good default (disabled logging)?

**Review by fotos — DISMISSED** [review]
:+1: LGTM 🚀 

I didn't check the `metrics.go` code thoroughly, yet from a cursory pass it looks reasonable to me.

File: README.md [hunk 4]
```diff
@@ -841,23 +841,26 @@ Kafka top-level configuration contains generic options that fit both publisher a
...
-| Topic    | Topic name                                                                             | `topic`       | `APP_PUBSUB_KAFKA_SUBSCRIBER_TOPIC`    | string | topic           |
-| Group ID | Client group id string. All clients sharing the same group.id belong to the same group | `group_id`    | `APP_PUBSUB_KAFKA_SUBSCRIBER_GROUP_ID` | string | service-name    |
+| Setting        | Description                                                                            | YAML variable     | Environment variable (ENV)                    | Type   | Possible Values |
+|----------------|----------------------------------------------------------------------------------------|-------------------|-----------------------------------------------|--------|-----------------|
+| Topic          | Topic name                                                                             | `topic`           | `APP_PUBSUB_KAFKA_SUBSCRIBER_TOPIC`           | string | topic           |
+| Group ID       | Client group id string. All clients sharing the same group.id belong to the same group | `group_id`        | `APP_PUBSUB_KAFKA_SUBSCRIBER_GROUP_ID`        | string | service-name    |
```
Comment (terranisu) [inline README.md:862]: > All clients sharing the same group.id belong to the same group

I'm not sure I understand this sentence. Did you mean `All clients are sharing the same group. Id belongs to...`?

**Review by terranisu — DISMISSED** [review]
LGTM 👍🏻  I left one comment though

File: README.md (hunk 4 above)
Comment (Neurostep) [reply README.md:862]: @terranisu I it taken from [the librdkafka documentation](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md)

File: README.md (hunk 2 above)
Comment (Neurostep) [reply README.md:844]: @fotos We don't have `broker` related configuration section atm, and I don't think we need it. We will adjust it in the future if we need one.

File: pkg/logger/kafka/kafka.go (hunk 3 above)
Comment (Neurostep) [reply pkg/logger/kafka/kafka.go:58]: @fotos Default logger is set to `error` [here](https://github.com/scribd/go-sdk/pull/42/files#diff-e769714803d020600fbaa166105b2afce5ed5bcc83c1c9a181944e48cad92e13R20). The code you are referring to is to handle the unknown log level provided.

Comment (fotos) [reply pkg/logger/kafka/kafka.go:58]: Got it, thanks for clarifying!

File: README.md (hunk 2 above)
Comment (fotos) [reply README.md:844]: Sounds good.

File: README.md (hunk 4 above)
Comment (terranisu) [reply README.md:864]: oh, I see now. The dot in `group.id` is what confused me. Can we remove it then? In the official documentation, the dot is part of the name of the parameter `group.id` so it's clear why we have `group.id` in the description. In our case, we don't need that dot, right?

Comment (Neurostep) [reply README.md]: @terranisu I have removed `.`

**Review by terranisu — DISMISSED** [review]
LGTM

File: README.md [hunk 5]
```diff
@@ -1114,6 +1119,117 @@ the SDK call stack. To learn more about these functions you can start by
 reading about them on [the AWS developer
 blog](https://aws.amazon.com/blogs/developer/v2-aws-sdk-for-go-adds-context-to-api-operations).
 
+### PubSub instrumentation and logging
+
+#### Kafka
+
+`go-sdk` provides a flexible way to instrument Kafka client by providing a logging, tracing and sending metrics to
+DataDog.
```
Comment (fotos) [inline README.md:1127]:
```suggestion
`go-sdk` provides a flexible way to instrument the Kafka client with logging, tracing and metrics capabilities.
```

As far as I can see this PR does not add anything DataDog specific. It users the SDK logger and metrics capabilities which (gladly) abstract away the underlying destination.

File: README.md [hunk 6]
```diff
@@ -1114,6 +1119,117 @@ the SDK call stack. To learn more about these functions you can start by
...
+
+[franz-go](https://github.com/twmb/franz-go/) Kafka client library is wrapped by the `go-sdk` and traces calls to Kafka.
```
Comment (fotos) [inline README.md:1129]:
```suggestion
The [franz-go](https://github.com/twmb/franz-go/) Kafka client library is wrapped by the `go-sdk` and traces calls to Kafka.
```

File: README.md [hunk 7]
```diff
@@ -1114,6 +1119,117 @@ the SDK call stack. To learn more about these functions you can start by
...
+
+For logging, `go-sdk` provides [franz-go compatible](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#WithLogger) logger which is basically a plugin
```
Comment (fotos) [inline README.md:1131]: nit:

```suggestion
For logging, `go-sdk` provides a [franz-go compatible](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#WithLogger) logger which is basically a plugin
```

File: README.md [hunk 8]
```diff
@@ -1114,6 +1119,117 @@ the SDK call stack. To learn more about these functions you can start by
...
+                logger,
+                sdkkafkalogger.WithLevel(sdklogger.Level(config.Logger.ConsoleLevel)),
+            ),
+        ),
+    }
+}
+```
+
+For metrics, `go-sdk` provides `franz-go` plugin to publish metrics to the DataDog via DogStatsd. It is implemented as
+[hooks](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#WithHooks). Metrics split into three categories: `broker`, `producer` and `consumer`.
```
Comment (fotos) [inline README.md:1154]: nit:

```suggestion
For metrics, `go-sdk` provides a `franz-go` plugin to publish metrics to DataDog via DogStatsd. It is implemented
as a [hook](https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#WithHooks). Metrics are split into three categories: `broker`, `producer` and `consumer`.
```

File: README.md [hunk 9]
```diff
@@ -1114,6 +1119,117 @@ the SDK call stack. To learn more about these functions you can start by
...
+```go
+import (
+    "github.com/twmb/franz-go/pkg/kgo"
+
+    sdkkafkamertics "github.com/scribd/go-sdk/pkg/metrics/kafka"
+)
+
+func main() {
+    kafkaClientOpts := []kgo.Opt{
+		// no sampling rate specified
```
Comment (fotos) [inline README.md:1217]: nit: something is off with indentation… probably tabs VS spaces.

**Review by fotos — DISMISSED** [review]
:+1: LGTM 🚀 

Great stuff @Neurostep! 👏 

Left a few nitpick comments for the documentation.

**Review by terranisu — APPROVED** [review]

**Review by fotos — APPROVED** [review]
:+1: LGTM 🚀

**Pull request merged**